eth/downloader/whitelist/checkpoint.go

```diff
@@ -0,0 +1,145 @@
+package whitelist
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type checkpoint struct {
+	m                sync.Mutex
+	checkpointHash   common.Hash // Whitelisted Checkpoint Hash, populated by reaching out to heimdall
+	checkpointNumber uint64      // Checkpoint Number , populated by reaching out to heimdall
+	interval         uint64      // Interval, until which we can allow importing
+	doExist          bool
+}
+
+var (
+	ErrCheckpointMismatch = errors.New("checkpoint mismatch")
+	ErrNoRemoteCheckoint  = errors.New("remote peer doesn't have a checkoint")
+)
+
+// IsValidPeer checks if the chain we're about to receive from a peer is valid or not
+// in terms of reorgs. We won't reorg beyond the last bor checkpoint submitted to mainchain.
+func (w *checkpoint) IsValidPeer(remoteHeader *types.Header, fetchHeadersByNumber func(number uint64, amount int, skip int, reverse bool) ([]*types.Header, []common.Hash, error)) (bool, error) {
+	// We want to validate the chain by comparing the last checkpointed block
+	// we're storing in `checkpointWhitelist` with the peer's block.
+	w.m.Lock()
+
+	// Check for availaibility of the last checkpointed block.
+	// doExist will be false if our heimdall is not responding
+	// or we're running without it.
+	if !w.doExist {
+		w.m.Unlock()
+		// worst case, we don't have the checkpoint in memory
+		return true, nil
+	}
+
+	// Fetch the last checkpoint entry
+	lastCheckpointBlockNum := w.checkpointNumber
+	lastCheckpointBlockHash := w.checkpointHash
+
+	w.m.Unlock()
+
+	// todo: we can extract this as an interface and mock as well or just test IsValidChain in isolation from downloader passing fake fetchHeadersByNumber functions
+	headers, hashes, err := fetchHeadersByNumber(lastCheckpointBlockNum, 1, 0, false)
+	if err != nil {
+		return false, fmt.Errorf("%w: last checkpoint %d, err %v", ErrNoRemoteCheckoint, lastCheckpointBlockNum, err)
+	}
+
+	if len(headers) == 0 {
+		return false, fmt.Errorf("%w: last checkpoint %d", ErrNoRemoteCheckoint, lastCheckpointBlockNum)
+	}
+
+	reqBlockNum := headers[0].Number.Uint64()
+	reqBlockHash := hashes[0]
+
+	// Check against the latest checkpointed block
+	if reqBlockNum == lastCheckpointBlockNum && reqBlockHash == lastCheckpointBlockHash {
+		return true, nil
+	}
+
+	return false, ErrCheckpointMismatch
+}
+
+// IsValidChain checks the validity of chain by comparing it
+// against the local checkpoint entry
+func (w *checkpoint) IsValidChain(currentHeader *types.Header, chain []*types.Header) bool {
+	// Check if we have checkpoint to validate incoming chain in memory
+	if !w.doExist {
+		// We don't have whitelisted checkpoint, no additional validation will be possible
+		return true
+	}
+
+	// Return if we've received empty chain
+	if len(chain) == 0 {
+		return false
+	}
+
+	w.m.Lock()
```

Comment:
What fields we're guarding?

Reply:
w.checkpointNumber, It might be possible that when we are reading this,  other routine might be writing to it.